libp2p/Host.cpp

```diff
@@ -145,70 +141,73 @@ void Host::stop()
+    // capabilities. Also enables reuse of the ioservice)
     m_ioService.reset();
-
-    DEV_GUARDED(x_timers)
-        m_timers.clear();
-    
-    // shutdown acceptor
-    m_tcp4Acceptor.cancel();
-    if (m_tcp4Acceptor.is_open())
-        m_tcp4Acceptor.close();
-
-    // There maybe an incoming connection which started but hasn't finished.
-    // Wait for acceptor to end itself instead of assuming it's complete.
-    // This helps ensure a peer isn't stopped at the same time it's starting
-    // and that socket for pending connection is closed.
-    while (m_accepting)
-        m_ioService.poll();
-
-    // stop capabilities (eth: stops syncing or block/tx broadcast)
-    for (auto const& h: m_capabilities)
-        h.second->onStopping();
-
-    // disconnect pending handshake, before peers, as a handshake may create a peer
-    for (unsigned n = 0;; n = 0)
-    {
-        DEV_GUARDED(x_connecting)
-            for (auto const& i: m_connecting)
-                if (auto h = i.lock())
-                {
-                    h->cancel();
-                    n++;
-                }
-        if (!n)
-            break;
-        m_ioService.poll();
-    }
-    
-    // disconnect peers
-    for (unsigned n = 0;; n = 0)
+    if (m_capabilities.size())
     {
-        DEV_RECURSIVE_GUARDED(x_sessions)
-            for (auto i: m_sessions)
-                if (auto p = i.second.lock())
-                    if (p->isConnected())
+        DEV_GUARDED(x_timers)
+            m_timers.clear();
+
+        // shutdown acceptor
+        m_tcp4Acceptor.cancel();
+        if (m_tcp4Acceptor.is_open())
+            m_tcp4Acceptor.close();
+
+        // There maybe an incoming connection which started but hasn't finished.
+        // Wait for acceptor to end itself instead of assuming it's complete.
+        // This helps ensure a peer isn't stopped at the same time it's starting
+        // and that socket for pending connection is closed.
+        while (m_accepting)
+            m_ioService.poll();
+
+        // stop capabilities (eth: stops syncing or block/tx broadcast)
+        for (auto const& h : m_capabilities)
+            h.second->onStopping();
+
+        // disconnect pending handshake, before peers, as a handshake may create a peer
+        for (unsigned n = 0;; n = 0)
+        {
+            DEV_GUARDED(x_connecting)
+                for (auto const& i : m_connecting)
+                    if (auto h = i.lock())
                     {
-                        p->disconnect(ClientQuit);
+                        h->cancel();
                         n++;
                     }
-        if (!n)
-            break;
+            if (!n)
+                break;
+            m_ioService.poll();
+        }
 
-        // poll so that peers send out disconnect packets
-        m_ioService.poll();
-    }
+        // disconnect peers
+        for (unsigned n = 0;; n = 0)
```

Comment:
WTF? Use `int` at least. `unsigned` is bad for a loop counter.

Reply:
This is the old code, only indentation changed; to me using the for loop itself is strange here; counter not needed at all, it could be a flag.

Reply:
Yes, I know. Use flag here. The rest should stay.

Reply:
Will replace this (and the code above it which disconnects pending handshakes) with a flag